Keycloak direct integration (#8411)
geosolutions-it/MapStore2: docs/developer-guide/integrations/users/keycloak.md

# Keycloak Integrations

## General

MapStore supports various functionalities to integrate with Keycloak:

- [**OpenID support**](../openId/#keycloak): Allows to login to MapStore using a keycloak account.
- [**Single sign on**](#single-sign-on-integration): Enhances the OpenID support by detecting a session in the keycloak realm and automatically login/logout from MapStore
- [**Direct user integration**](#direct-user-integration): Enhances the OpenID support making MapStore use keycloak as unique Identity Manager System (IdM), replacing the MapStore DB with Keycloak REST API.

## OpenID

> **Keycloak OpenID support** allows to use a keycloak instance as Identity Provider (IdP) via OpenID Connect (OIDC), so the user can login to MapStore using an existing account in keycloak.

Reviewer: **Keycloak OpenID support** allows to use a keycloak instance as Identity Provider (IdP) via OpenID Connect (OIDC), so that the user can login to MapStore using an existing account in keycloak.

Author: ```suggestion
**Keycloak OpenID support** allows to use a keycloak instance as Identity Provider (IdP) via OpenID Connect (OIDC), so that the user can login to MapStore using an existing account in keycloak.
```

<img src="../img/kc-create-mapstore-client.jpg" class="ms-docimage"  style="max-width:500px;"/>

- Configure it as `Public`
- Insert in  **"Valid Redirect URIs"** your MapStore base root, with a `*` at the end (e.g. `https://my.mapstore.site.com/mapstore/*`)
- Insert in **"Web Origins"** your MapStore base domain name. (e.g. `https://my.mapstore.site.com`)

<img src="../img/kc-configure-mapstore-client.jpg" class="ms-docimage"  style="max-width:500px;"/>

- Click on Save button, then open the *Installation* tab, select the `Keycloak OIDC JSON` format, and copy the JSON displayed below.

<img src="../img/copy-mapstore-client-config.jpg" class="ms-docimage"  style="max-width:500px;"/>

### Configure SSO in MapStore

After configuring the open [openID integration](../openId#keycloak), you will have an entry named `keycloak` in `authenticationProviders`.
In this entry, you will have to add `"sso":{"type":"keycloak"}` and `config: "<configuration coped from keycloak>"`.

e.g.

```json
{
       "authenticationProviders": [
      {
        "type": "openID",
        "provider": "keycloak",
        "config": {
          "realm": "master",

>           "auth-server-url": "http://localhost:8080/",

Author: This is only an example. The guide says to copy and pasted

>           "auth-server-url": "http://localhost:8080/",

Reviewer: ok fine

          "ssl-required": "external",
          "resource": "mapstore-client",
          "public-client": true,
          "confidential-port": 0
        },
        "sso": {
          "type": "keycloak"
        }
      }
    ],
}
```

Here implementation details about [keycloak login workflow](../keycloak-sso-impl/).

## Direct user integration

By default MapStore can integrate openID login with Keycloak and also supports integration with Keycloak SSO.

> By default users that do the login with Keycloak are created on the database.

Reviewer: Here I would also specify that the database in question is the geostore database that stores MapStore users' credentials and roles.

Author: ```suggestion
By default users that do the login with Keycloak are created on the MapStore's database, where all the users and user-groups are normally stored.
```

> By default users that do the login with Keycloak are created on the database.

Author: ```suggestion
By default users that login with Keycloak are created on the database.
```

> Anyway MapStore can interact with Keycloak REST API to provide a direct integration without persisting anything on the database.

Reviewer: Here I would also specify that  'the database' is the geostore database.

Author: ```suggestion
Anyway MapStore can interact with Keycloak REST API to provide a direct integration without persisting anything on the MapStore's database.
```

> This provides a more strict integration between the applications, allowing to assign roles and groups directly from keycloak, and avoiding any synchronization issue.

Author: ```suggestion
This provides a stricter integration between the applications, allowing the assignment of roles and groups directly from keycloak, and avoiding any synchronization issue.
```

This is actually the only way to map keycloak roles to MapStore groups.

In this scenario the integration MapStore replaces the user and user-group database tables with the keycloak REST API.

!!! note

>     This integration denies the possibility to use the user database and to use the admin user interface to manage users and groups.

Author: ```suggestion
    This integration prevents a MapStore user with role `ADMIN`  from using the user database and from accessing the users/groups management page.
```

>     This integration denies the possibility to use the user database and to use the admin user interface to manage users and groups.

Author: Not properly, let me explain it better.
```suggestion
    This integration denies the possibility to write the user database also any writing operation using the REST API is not supported yet.
```

    For this reason we suggest to disable the `UserManager`, `GroupManager` plugins, and remove the `authenticationProviders` entry of type `geostore`, if any, because the standard login with username and password is not allowed for the db users.
    In case of integration with GeoServer, also GeoServer should be connected to Keycloak for users, and not to the MapStore database.

### Configure direct integration with keycloak

> In order to apply the direct integration with keycloak you will have to:

Reviewer: To enable the direct integration with keycloak you will have to:

By default MapStore can integrate openID login with Keycloak and also supports integration with Keycloak SSO.

By default users that login with Keycloak are created on the database.
Anyway MapStore can interact with Keycloak REST API to provide a direct integration without persisting anything on the MapStore's database.
This provides a stricter integration between the applications, allowing the assignment of roles and groups directly from keycloak, and avoiding any synchronization issue.

This is actually the only way to map keycloak roles to MapStore groups.

In this scenario the integration MapStore replaces the user and user-group database tables with the keycloak REST API.

!!! note

>     This integration denies the possibility to write the user database also any writing operation using the REST API is not supported yet.

Reviewer: ```suggestion
    This integration disables reading and writing to the users' and groups' database and replaces it with the Keycloak REST API, with read-only support.
```
